### PR TITLE
XCOMMONS-2068: Possible OutOfBoundsException when performing diff

### PR DESCRIPTION
**Pull request by surli**
## JIRA

[XCOMMONS-2068](https://jira.xwiki.org/browse/XCOMMONS-2068): Possible OutOfBoundsException when performing diff
[XCOMMONS-2137](https://jira.xwiki.org/browse/XCOMMONS-2137): Possible IllegalArgumentException when performing a merge

##  Description

This changes provides new integration tests that helped investigating
the problems in the diff algorithm. Basically the main issue was in the
creation of conflicts information: all subsequent errors are most likely
coming from there.
The changes here are mostly about computing a proper index for extract
conflict information, and actually enforcing proper usage of sublist API
to avoid problems: this might displays some loggers if other problems
occurs in the algorithm, but this would not break the caller.

Note that the integration tests file are produced using https://gist.github.com/surli/5506e2427daa4cad6279fbfea8f7f172

### CODE REVIEW

File: xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-display/src/test/java/org/xwiki/diff/display/internal/DefaultUnifiedDiffDisplayerTest.java [hunk 1]
```diff
@@ -533,4 +533,19 @@ private static String readContent(String fileName) throws IOException
...
+        List<String> previous = readLines("integration3/previous.txt");
+        List<String> current = readLines("integration3/current.txt");
+        List<String> next = readLines("integration3/next.txt");
+        MergeResult<String> mergeResult = getDiffManager().merge(previous, next, current, null);
+        assertEquals(3, mergeResult.getConflicts().size());
+
+        DiffResult<String> diffResult = getDiffManager().diff(previous, current, null);
+        List<UnifiedDiffBlock<String, Object>> unifiedDiffBlocks = unifiedDiffDisplayer.display(diffResult);
+
+        unifiedDiffBlocks = unifiedDiffDisplayer.display(diffResult, mergeResult.getConflicts());
```
Comment (manuelleduc) [inline xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-display/src/test/java/org/xwiki/diff/display/internal/DefaultUnifiedDiffDisplayerTest.java:549]: Is something missing? Just wondering why this variable is affected but never used afterward.

File: xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-display/src/main/java/org/xwiki/diff/display/internal/DefaultUnifiedDiffDisplayer.java [hunk 2]
```diff
@@ -341,13 +352,14 @@ public void setLastConflict(Conflict<E> lastConflict)
     private <E> boolean canDeltaSplitted(Conflict<E> conflict, Delta<E> delta)
     {
         if (conflict != null) {
-            // a delta can be splitted only if one of its chunk is > 1
+            // a delta can be splitted only if it concerns a change
```
Comment (manuelleduc) [inline xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-display/src/main/java/org/xwiki/diff/display/internal/DefaultUnifiedDiffDisplayer.java:355]: Is the old comment `a delta can be splitted only if one of its chunk is > 1` deprecated? Reading the code it looks like it is still relevant (in addition to your new comment).

PS: split is an irregular word.

File: xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-api/src/main/java/org/xwiki/diff/internal/DefaultDiffManager.java [hunk 3]
```diff
@@ -494,19 +518,25 @@
...
             }
-            subsetPrevious = new ArrayList<>(previous.subList(newIndex, newOffsetEnd));
+            subsetPrevious = new ArrayList<>(extractFromList(previous, newIndex, newOffsetEnd));
         }
 
-        List<E> subsetNext = new ArrayList<>(extractConflictPart(deltaNext, previous, next, chunkSize, index));
-        List<E> subsetCurrent = new ArrayList<>(extractConflictPart(deltaCurrent, previous, current, chunkSize, index));
+        // We need to always use the minimum index to extract the conflict.
+        newIndex = Math.min(prevMinIndex, nextMinIndex);
+        List<E> subsetNext = new ArrayList<>(extractConflictPart(deltaNext, previous, next, chunkSize, newIndex));
```
Comment (manuelleduc) [inline xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-api/src/main/java/org/xwiki/diff/internal/DefaultDiffManager.java:537]: Could the wrapping in a new list be moved inside extractConflictPart? Seems like it is needed every time this method is called, and it looks error prone.

File: xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-display/src/main/java/org/xwiki/diff/display/internal/DefaultUnifiedDiffDisplayer.java (hunk 2 above)
Comment (surli) [reply xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-display/src/main/java/org/xwiki/diff/display/internal/DefaultUnifiedDiffDisplayer.java]: Indeed, thanks. Also fixed the typo.

File: xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-api/src/main/java/org/xwiki/diff/internal/DefaultDiffManager.java (hunk 3 above)
Comment (surli) [reply xwiki-commons-core/xwiki-commons-diff/xwiki-commons-diff-api/src/main/java/org/xwiki/diff/internal/DefaultDiffManager.java]: Indeed I improved things.

**Pull request merged**